### PR TITLE
feat: Added support for binding ports based on the USB device location

### PR DESCRIPTION
**Pull request by alekseiapa**
- Added support for binding ports based on the location of a USB device

Closes https://github.com/espressif/pytest-embedded/issues/118

### CODE REVIEW

File: pytest-embedded-serial-esp/pytest_embedded_serial_esp/serial.py [hunk 1]
```diff
@@ -27,6 +27,7 @@ def __init__(
         target: Optional[str] = None,
         beta_target: Optional[str] = None,
         port: Optional[str] = None,
+        port_location: Optional[str] = None,
```
Comment (hfudev) [inline pytest-embedded-serial-esp/pytest_embedded_serial_esp/serial.py:30]: no need here.

File: pytest-embedded-serial/pytest_embedded_serial/serial.py [hunk 2]
```diff
@@ -52,12 +54,26 @@ def __init__(
...
+            for port in list_ports.comports():
+                if port.location == port_location:
+                    if self.port and port.device != self.port:
+                        is_port_found = False
+                        break
+                    self.port = port.device
+                    is_port_found = True
+                    break
+            if not is_port_found:
+                raise ValueError('The specified port cannot be found in the provided location.')
```
Comment (hfudev) [inline pytest-embedded-serial/pytest_embedded_serial/serial.py:74]: the error message should be more friendly. and you could use the for...else... to simplify it a bit.

```python
        if port_location:
            for port in list_ports.comports():
                if port.location == port_location:
                    if self.port and port.device != self.port:
                        raise ValueError(f'The specified location {port_location} binds with port {port.device}, not {self.port}')

                    self.port = port.device
                    break
            else:
                raise ValueError(f'The specified location {port_location} cannot be found.')
```

File: pytest-embedded-serial/pytest_embedded_serial/serial.py [hunk 3]
```diff
@@ -52,12 +54,26 @@ def __init__(
         self._meta = meta
 
         self.port = port
+        self.port_location = port_location
```
Comment (hfudev) [inline pytest-embedded-serial/pytest_embedded_serial/serial.py:57]: doesn't need this attribute

**Review by hfudev — CHANGES_REQUESTED** [review]
LGTM! Please also add the fixture `port_location` to let users define it in `pytest.mark.parametrize`, like the other cli options

File: pytest-embedded-serial-esp/pytest_embedded_serial_esp/serial.py (hunk 1 above)
Comment (hfudev) [reply pytest-embedded-serial-esp/pytest_embedded_serial_esp/serial.py:30]: The kwargs that is not been captured in this subclass `EspSerial` would pass to `Serial` here: https://github.com/espressif/pytest-embedded/blob/main/pytest-embedded-serial-esp/pytest_embedded_serial_esp/serial.py#L117

**Comment by alekseiapa** [comment]
@hfudev Could you please look over the changes one more time ? Thank you!

File: pytest-embedded-serial/pytest_embedded_serial/serial.py [hunk 4]
```diff
@@ -54,10 +55,24 @@ def __init__(
         self.port = port
         self.baud = baud
 
-        if port is None:
-            raise ValueError('Please specify port')
+        port_location = kwargs.get('port_location', None)
```
Comment (hfudev) [inline pytest-embedded-serial/pytest_embedded_serial/serial.py:58]: Here the `port_location` from param or cli is not actually been passed. Should be passed here: https://github.com/espressif/pytest-embedded/blob/main/pytest-embedded/pytest_embedded/plugin.py#L859

You may refer to other cli options/fixtures as examples

File: pytest-embedded/pytest_embedded/plugin.py [hunk 5]
```diff
@@ -115,6 +115,7 @@ def pytest_addoption(parser):
 
     serial_group = parser.getgroup('embedded-serial')
     serial_group.addoption('--port', help='serial port. (Env: "ESPPORT" if service "esp" specified, Default: "None")')
+    serial_group.addoption('--port-location', help='USB device location string ("<bus>-<port>[-<port>]…")')
```
Comment (hfudev) [inline pytest-embedded/pytest_embedded/plugin.py:118]:
```suggestion
    serial_group.addoption('--port-location', help='USB device location string ("<bus>-<port>[-<port>]…"). Default: None')
```

**Review by hfudev — CHANGES_REQUESTED** [review]

File: pytest-embedded/pytest_embedded/plugin.py [hunk 6]
```diff
@@ -988,6 +999,7 @@ def _fixture_classes_and_options(
                 kwargs[fixture] = {
                     'msg_queue': msg_queue,
                     'port': port,
+                    'port_location': port_location,
```
Comment (hfudev) [inline pytest-embedded/pytest_embedded/plugin.py:1002]: also need to add this line here: https://github.com/espressif/pytest-embedded/pull/141/files#diff-b5a6b0a6e20cb69a3ce927b538961a62b3a189870fb359992536e7cc2618cfa5R966

**Comment by alekseiapa** [comment]
Sorry for the inadvertence. There are a quite lot of things that are new to me.

**Comment by alekseiapa** [comment]
> @alekseiapa https://github.com/espressif/pytest-embedded/actions/runs/3218060372/jobs/5261770866#step:5:786 seems not passing the CI.

It should be solved now

**Comment by alekseiapa** [comment]
> > @alekseiapa https://github.com/espressif/pytest-embedded/actions/runs/3218060372/jobs/5261770866#step:5:786 seems not passing the CI.
> 
> It should be solved now

@hfudev finally passed CI

**Comment by hfudev** [comment]
@alekseiapa Thanks for your contribution!

**Pull request merged**